library/std/src/sys/thread_local/guard/key.rs

```diff
@@ -7,17 +7,38 @@ use crate::sys::thread_local::key::{LazyKey, set};
 
 #[cfg(target_thread_local)]
 pub fn enable() {
-    use crate::sys::thread_local::destructors;
+    fn enable_thread() {
+        static DTORS: LazyKey = LazyKey::new(Some(run_thread));
 
-    static DTORS: LazyKey = LazyKey::new(Some(run));
+        // Setting the key value to something other than NULL will result in the
+        // destructor being run at thread exit.
+        unsafe {
+            set(DTORS.force(), ptr::without_provenance_mut(1));
+        }
+
+        unsafe extern "C" fn run_thread(_: *mut u8) {
+            run()
+        }
+    }
+
+    #[cfg(target_has_atomic_load_store = "8")]
+    fn enable_process() {
+        use crate::sync::atomic::{AtomicBool, Ordering};
+        use crate::sys::thread_local::key::at_process_exit;
 
-    // Setting the key value to something other than NULL will result in the
-    // destructor being run at thread exit.
-    unsafe {
-        set(DTORS.force(), ptr::without_provenance_mut(1));
+        static REGISTERED: AtomicBool = AtomicBool::new(false);
+        if !REGISTERED.swap(true, Ordering::AcqRel) {
```

Comment:
Relaxed memory ordering is enough for this sort of thing.

Reply:
My fear of memory models always makes me choose the safest option ;-)